File: Hardware/ESP-Server/MQTT-communication.cpp
#include "esp32-hal-gpio.h"
#include "MQTT-communication.h"
#include "BLE-Server.h"
#include <WiFi.h>
#include <PubSubClient.h>

const char* ssid = "OPTELA";
const char* password = "";
const char* mqtt_server = "93.155.224.232";
const int port = 5728;

WiFiClient espClient;
PubSubClient client(espClient);

#define mqttACControl "ac/control"
#define mqttConnectionWifi "connection/wifi"

void topicsSubscribe() {
  client.subscribe(mqttACControl, 1);
  client.subscribe(mqttConnectionWifi, 1);
}

void callback(char* topic, byte* message, unsigned int length) {
  String receivedMessage;
  for (int i = 0; i < length; i++) {
    receivedMessage += (char)message[i];
  }

  if (String(topic) == mqttACControl) {
    Serial.println("Message received from ac/control");
    if (receivedMessage == "TURN_ON") {
      client.publish("ac/report", "Turn ON signal sent to AC");
      flashLED();
    } else if (receivedMessage == "TURN_OFF") {
      client.publish("ac/report", "Turn OFF signal sent to AC");
      flashLED();
    } else if (receivedMessage == "TURN_LED_ON") {
      digitalWrite(2, HIGH);
      delay(3000);
    } else if (receivedMessage == "TURN_LED_OFF") {
      digitalWrite(2, LOW);
      flashLED();
    } else {
      client.publish("ac/report", "Unknown command sent to ac/control");
      flashLED();
    }
  }

  if (String(topic) == mqttConnectionWifi) {
    Serial.println("Message received from connection/wifi");
    int delimeterIndex = receivedMessage.indexOf('/');
    if (delimeterIndex != -1) {
      String newSSID = receivedMessage.substring(0, delimeterIndex);
      String newPass = receivedMessage.substring(delimeterIndex + 1);
      Serial.println(newSSID);
      Serial.println(newPass);
      client.publish("ac/report", "Changed WiFi connection");
      connectWifi(newSSID, newPass);  // TODO: If new connection doesn't exist stay in previous connection
      flashLED();
    } else {
      client.publish("ac/report", "Couldn't change WiFi because it wasn't sent in correct format");
    }
  }

  Serial.println("Response sent to ac/report");
  delay(1000);
}

void mqtt_setup() {
  pinMode(2, OUTPUT);
  // connectWifi(ssid, password);  // currently handled by the BLE Server
  client.setServer(mqtt_server, port);
  client.setCallback(callback);
}

void reconnect() {
  if (client.connect("ESP32Client", "ac/control", 1, true, "Hitler")) {
    Serial.println("Connected");
    Serial.println();
    topicsSubscribe();
  } else {
    delay(5000);
  }
  //topicsSubscribe();
}

void flashLED() {
  digitalWrite(2, HIGH);
  delay(100);
  digitalWrite(2, LOW);
  delay(50);
  digitalWrite(2, HIGH);
  delay(100);
  digitalWrite(2, LOW);
  delay(1000);
}

void mqtt_loop_step() {
  if (!client.connected()) {
    reconnect();
  }
  client.loop();
}

File: Back-End/mqtt-node/mqtt-communication.js
import mqtt from 'mqtt';

const protocol = 'mqtt'
const host = '93.155.224.232'
const port = '5728'

const connectURL = `${protocol}://${host}:${port}`

// const responseTopic = 'ac/report'
const controlTopic = 'ac/control'
const wifiConnectionTopic = 'connection/wifi'

class ESPClient {
    constructor(options) {
        this.client = mqtt.connect(connectURL, options)
        this.clientId = options.clientId
        this.subscribedTopics = new Set();

        this.client.on('connect', () => {
            console.log(`[${this.clientId}] Connected`)
        })

        this.client.on('error', () => {
            console.log(`[${this.clientId}] Couldn't connect because of an error`)
        })

        this.client.on('message', (topic, message) => {
            if(this.subscribedTopics.has(topic)) {
                console.log(`[${this.clientId}] Message received on topic '${topic}': ${message}`)
            }
        })

        this.client.on('disconnect', () => {
            console.log(`[${this.clientId}] Disconnected`)
        })
    }

    publish(topic, message) {
        this.client.publish(topic, message, (error) => {
            if (error) {
                console.log(`[${this.clientId}] Couldn't publish`)
            } else {
                console.log(`[${this.clientId}] Published to ${topic}`)
            }
        })

    }

    subscribe(topic) {
        this.client.subscribe(topic, (error) => {
            if (error) {
                console.error(`[${this.clientId}] Failed to subscribe to topic`, error);
            } else {
                console.log(`[${this.clientId}] Subscribed to ${topic}`)
                this.subscribedTopics.add(topic)
            }
        })
    }

    turnLEDOn() {
        this.client.publish(controlTopic, 'TURN_LED_ON')
        console.log(`[${this.clientId}] Sent TURN_LED_ON command`)
    }

    turnLEDOff() {
        this.client.publish(controlTopic, 'TURN_LED_OFF')
        console.log(`[${this.clientId}] Sent TURN_LED_OFF command`)
    }

    changeWifi(ssid, pass) {
        this.client.publish(wifiConnectionTopic, `${ssid}/${pass}`)
        console.log(`[${this.clientId}] Sent updated WiFi config`)
    }
}

const esp = new ESPClient({
    clientId: 'mqtt_pc1', // we can use UUID
    clean: false,
    connectTimeout: 4000,
    reconnectPeriod: 1000,
})

esp.turnLEDOff()
esp.changeWifi('OPTELA', '')